File: common/src/main/java/juuxel/adorn/block/SofaBlock.java
package juuxel.adorn.block;

import juuxel.adorn.block.property.FrontConnection;
import juuxel.adorn.lib.AdornStats;
import juuxel.adorn.lib.AdornTags;
import juuxel.adorn.util.ShapeRotation;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.component.DataComponents;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.Identifier;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.util.RandomSource;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.attribute.BedRule;
import net.minecraft.world.attribute.EnvironmentAttributes;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.ScheduledTickAccess;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.level.block.SimpleWaterloggedBlock;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Fluids;
import net.minecraft.world.level.pathfinder.PathComputationType;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jspecify.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class SofaBlock extends SeatBlock implements SimpleWaterloggedBlock, SneakClickHandler, BlockWithDescription {
    public static final EnumProperty<Direction> FACING = BlockStateProperties.HORIZONTAL_FACING;
    public static final BooleanProperty CONNECTED_LEFT = BooleanProperty.create("connected_left");
    public static final BooleanProperty CONNECTED_RIGHT = BooleanProperty.create("connected_right");
    public static final EnumProperty<FrontConnection> FRONT_CONNECTION = EnumProperty.create("front", FrontConnection.class);
    public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;

    private static final VoxelShape[] OUTLINE_SHAPES = buildShapes(false);
    private static final VoxelShape[] COLLISION_SHAPES = buildShapes(true);
    private static final String DESCRIPTION_KEY = "block.adorn.sofa.description";

    public SofaBlock(Properties settings) {
        super(settings);
        registerDefaultState(defaultBlockState()
            .setValue(FRONT_CONNECTION, FrontConnection.NONE)
            .setValue(CONNECTED_LEFT, false)
            .setValue(CONNECTED_RIGHT, false)
            .setValue(WATERLOGGED, false));
    }

    @Override
    public Identifier getSittingStat() {
        return AdornStats.SIT_ON_SOFA;
    }

    @Override
    public String getDescriptionKey() {
        return DESCRIPTION_KEY;
    }

    @Override
    protected InteractionResult useItemOn(ItemStack stack, BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
        if (stack.is(AdornTags.FURNITURE_DYES)) {
            var dye = stack.get(DataComponents.DYE);
            if (dye != null) {
                world.setBlockAndUpdate(pos, AdornBlocks.SOFAS.pick(dye).get().withPropertiesOf(state));
                world.playSound(player, pos, SoundEvents.WOOL_PLACE, SoundSource.BLOCKS, 1f, 0.8f);
                if (!player.getAbilities().instabuild) stack.shrink(1);
                if (!world.isClientSide()) player.awardStat(AdornStats.DYE_SOFA);
                return InteractionResult.SUCCESS;
            }
        }

        return super.useItemOn(stack, state, world, pos, player, hand, hit);
    }

    @Override
    public InteractionResult onSneakClick(BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hitResult) {
        var sleepingDirection = getSleepingDirection(world, pos);

        if (state.getValue(OCCUPIED)) {
            player.sendOverlayMessage(Component.translatable("block.adorn.sofa.occupied"));
            return InteractionResult.SUCCESS;
        }

        var bedRule = modifyBedRuleForSofas(
            world.environmentAttributes().getValue(EnvironmentAttributes.BED_RULE, pos)
        );
        if (bedRule.canSleep(world) && sleepingDirection != null) {
            if (!world.isClientSide()) {
                player.startSleepInBed(pos).ifLeft(reason -> {
                    if (reason.message() != null) {
                        player.sendOverlayMessage(reason.message());
                    }
                });
            }

            return InteractionResult.SUCCESS;
        } else {
            return InteractionResult.PASS;
        }
    }

    public static BedRule modifyBedRuleForSofas(BedRule rule) {
        return rule.canSleep() == BedRule.Rule.WHEN_DARK ? new BedRule(BedRule.Rule.ALWAYS, rule.canSetSpawn(), rule.explodes(), rule.errorMessage()) : rule;
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
        builder.add(FACING, CONNECTED_LEFT, CONNECTED_RIGHT, FRONT_CONNECTION, WATERLOGGED);
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        return updateConnections(
            defaultBlockState().setValue(FACING, ctx.getHorizontalDirection().getOpposite())
                .setValue(WATERLOGGED, ctx.getLevel().getFluidState(ctx.getClickedPos()).getType() == Fluids.WATER),
            ctx.getLevel(),
            ctx.getClickedPos()
        );
    }

    @Override
    public FluidState getFluidState(BlockState state) {
        return state.getValue(WATERLOGGED) ? Fluids.WATER.getSource(false) : super.getFluidState(state);
    }

    @Override
    protected BlockState updateShape(BlockState state, LevelReader world, ScheduledTickAccess tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, RandomSource random) {
        if (state.getValue(WATERLOGGED)) {
            tickView.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(world));
        }

        return updateConnections(state, world, pos);
    }

    private BlockState updateConnections(BlockState state, BlockGetter world, BlockPos pos) {
        var direction = state.getValue(FACING);
        var leftState = world.getBlockState(pos.relative(direction.getClockWise()));
        var rightState = world.getBlockState(pos.relative(direction.getCounterClockWise()));
        var frontState = world.getBlockState(pos.relative(direction));

        var connectedLeft = leftState.getBlock() instanceof SofaBlock && (leftState.getValue(FACING) == direction || (leftState.getValue(FACING) == direction.getCounterClockWise() && leftState.getValue(FRONT_CONNECTION) != FrontConnection.NONE));
        var connectedRight = rightState.getBlock() instanceof SofaBlock && (rightState.getValue(FACING) == direction || (rightState.getValue(FACING) == direction.getClockWise() && rightState.getValue(FRONT_CONNECTION) != FrontConnection.NONE));
        var connectedFront = frontState.getBlock() instanceof SofaBlock;
        var connectedFrontLeft = connectedFront && !connectedLeft && frontState.getValue(FACING) == direction.getCounterClockWise();
        var connectedFrontRight = connectedFront && !connectedRight && frontState.getValue(FACING) == direction.getClockWise();
        var frontConnection = FrontConnection.NONE;
        if (connectedFrontLeft) {
            frontConnection = FrontConnection.LEFT;
        } else if (connectedFrontRight) {
            frontConnection = FrontConnection.RIGHT;
        }

        return state
            .setValue(CONNECTED_LEFT, connectedLeft)
            .setValue(CONNECTED_RIGHT, connectedRight)
            .setValue(FRONT_CONNECTION, frontConnection);
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
        return OUTLINE_SHAPES[
            getShapeKey(
                state.getValue(FACING),
                state.getValue(CONNECTED_LEFT),
                state.getValue(CONNECTED_RIGHT),
                state.getValue(FRONT_CONNECTION)
            )
        ];
    }

    @Override
    public VoxelShape getCollisionShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
        return COLLISION_SHAPES[
            getShapeKey(
                state.getValue(FACING),
                state.getValue(CONNECTED_LEFT),
                state.getValue(CONNECTED_RIGHT),
                state.getValue(FRONT_CONNECTION)
            )
        ];
    }

    @Override
    public BlockState mirror(BlockState state, Mirror mirror) {
        return state.rotate(mirror.getRotation(state.getValue(FACING)));
    }

    @Override
    public BlockState rotate(BlockState state, Rotation rotation) {
        return state.setValue(FACING, rotation.rotate(state.getValue(FACING)));
    }

    @Override
    public boolean isPathfindable(BlockState state, PathComputationType type) {
        return false;
    }

    @Override
    public double getSittingOffset(Level world, BlockState state, BlockPos pos) {
        return 0.4375; // 7/16
    }

    @Override
    public Direction getPreferredDismountDirection(BlockState state, Entity passenger) {
        return state.getValue(FACING);
    }

    private static VoxelShape[] buildShapes(boolean thin) {
        var bottom = box(0.0, 2.0, 0.0, 16.0, 7.0, 16.0);
        var leftArms = ShapeRotation.buildShapeRotations(5, 7, 13, 16, 13, 16);
        var rightArms = ShapeRotation.buildShapeRotations(5, 7, 0, 16, 13, 3);
        var thinLeftArms = ShapeRotation.buildShapeRotations(5, 7, 14, 16, 13, 16);
        var thinRightArms = ShapeRotation.buildShapeRotations(5, 7, 0, 16, 13, 2);
        var backs = ShapeRotation.buildShapeRotations(0, 7, 0, 5, 16, 16);
        var leftCorners = ShapeRotation.buildShapeRotations(5, 7, 11, 16, 16, 16);
        var rightCorners = ShapeRotation.buildShapeRotations(5, 7, 0, 16, 16, 5);
        var booleans = new boolean[] { true, false };
        var result = new VoxelShape[48];
        for (var facing : FACING.getPossibleValues()) {
            for (var left : booleans) {
                for (var right : booleans) {
                    for (var front : FRONT_CONNECTION.getPossibleValues()) {
                        List<VoxelShape> parts = new ArrayList<>();
                        parts.add(backs.get(facing));

                        if (!left && front == FrontConnection.NONE) {
                            parts.add(thin ? thinLeftArms.get(facing) : leftArms.get(facing));
                        }
                        if (!right && front == FrontConnection.NONE) {
                            parts.add(thin ? thinRightArms.get(facing) : rightArms.get(facing));
                        }

                        switch (front) {
                            case LEFT -> parts.add(leftCorners.get(facing));
                            case RIGHT -> parts.add(rightCorners.get(facing));
                        }

                        int key = getShapeKey(facing, left, right, front);
                        var shape = Shapes.or(bottom, parts.toArray(VoxelShape[]::new));
                        result[key] = shape;
                    }
                }
            }
        }
        return result;
    }

    private static int getShapeKey(Direction facing, boolean left, boolean right, FrontConnection front) {
        return front.ordinal() << 4 | (left ? 1 : 0) << 3 | (right ? 1 : 0) << 2 | facing.get2DDataValue();
    }

    public static @Nullable Direction getSleepingDirection(BlockGetter world, BlockPos pos) {
        return getSleepingDirection(world, pos, false);
    }

    // TODO: Is the ignoreNeighbors = false state ever needed??
    public static @Nullable Direction getSleepingDirection(BlockGetter world, BlockPos pos, boolean ignoreNeighbors) {
        var state = world.getBlockState(pos);
        if (!(state.getBlock() instanceof SofaBlock)) return null;

        boolean connectedLeft = state.getValue(CONNECTED_LEFT);
        boolean connectedRight = state.getValue(CONNECTED_RIGHT);
        var frontConnection = state.getValue(FRONT_CONNECTION);
        var facing = state.getValue(FACING);

        if ((!connectedLeft && !connectedRight && frontConnection == FrontConnection.NONE) || (!ignoreNeighbors && state.getValue(OCCUPIED))) {
            return null;
        }

        Direction result;
        if (frontConnection != FrontConnection.NONE) {
            result = facing;
        } else if (connectedLeft) {
            result = facing.getClockWise();
        } else if (connectedRight) {
            result = facing.getCounterClockWise();
        } else {
            result = null;
        }

        if (result != null) {
            if (ignoreNeighbors) {
                return result;
            }
            var neighborState = world.getBlockState(pos.relative(result));
            if (neighborState.getBlock() instanceof SofaBlock && !neighborState.getValue(OCCUPIED)) {
                return result;
            }
        }

        return null;
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/SofaBlock.kt
@file:Suppress("DEPRECATION")
package juuxel.adorn.block

import com.google.common.collect.Sets
import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap
import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap
import juuxel.adorn.api.block.BlockVariant
import juuxel.adorn.block.property.FrontConnection
import juuxel.adorn.platform.PlatformBridges
import juuxel.adorn.util.buildShapeRotations
import juuxel.adorn.util.withBlock
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.ShapeContext
import net.minecraft.block.Waterloggable
import net.minecraft.entity.ai.pathing.NavigationType
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.fluid.Fluids
import net.minecraft.item.DyeItem
import net.minecraft.item.ItemPlacementContext
import net.minecraft.server.world.ServerWorld
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvents
import net.minecraft.state.StateManager
import net.minecraft.state.property.BooleanProperty
import net.minecraft.state.property.EnumProperty
import net.minecraft.state.property.Properties
import net.minecraft.util.ActionResult
import net.minecraft.util.BlockMirror
import net.minecraft.util.BlockRotation
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.BlockView
import net.minecraft.world.World
import net.minecraft.world.WorldAccess

class SofaBlock(variant: BlockVariant) :
    SeatBlock(variant.createSettings()),
    Waterloggable,
    SneakClickHandler {
    init {
        defaultState = defaultState
            .with(FRONT_CONNECTION, FrontConnection.NONE)
            .with(CONNECTED_LEFT, false)
            .with(CONNECTED_RIGHT, false)
            .with(WATERLOGGED, false)
    }

    override fun onUse(
        state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, hit: BlockHitResult
    ): ActionResult {
        val stack = player.getStackInHand(hand)
        val item = stack.item
        if (item is DyeItem) {
            world.setBlockState(pos, state.withBlock(AdornBlocks.SOFAS[item.color]!!))
            world.playSound(player, pos, SoundEvents.BLOCK_WOOL_PLACE, SoundCategory.BLOCKS, 1f, 0.8f)
            if (!player.abilities.creativeMode) stack.decrement(1)
            return ActionResult.SUCCESS
        }

        return super.onUse(state, world, pos, player, hand, hit)
    }

    override fun onSneakClick(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, hitResult: BlockHitResult): ActionResult {
        val sleepingDirection = getSleepingDirection(world, pos)
        return if (world.dimension.isNatural && sleepingDirection != null && !state[OCCUPIED]) {
            if (world is ServerWorld) {
                with(PlatformBridges.entities) {
                    player.trySleep(pos) {
                        world.setBlockState(pos, state.with(OCCUPIED, true))
                        val neighborPos = pos.offset(sleepingDirection)
                        world.setBlockState(neighborPos, world.getBlockState(neighborPos).with(OCCUPIED, true))
                        world.updateSleepingPlayers()
                    }
                }
            }
            ActionResult.SUCCESS
        } else ActionResult.PASS
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        super.appendProperties(builder)
        builder.add(FACING, CONNECTED_LEFT, CONNECTED_RIGHT, FRONT_CONNECTION, WATERLOGGED)
    }

    override fun getPlacementState(context: ItemPlacementContext): BlockState {
        return updateConnections(
            super.getPlacementState(context)!!.with(FACING, context.playerFacing.opposite)
                .with(WATERLOGGED, context.world.getFluidState(context.blockPos).fluid == Fluids.WATER),
            context.world,
            context.blockPos
        )
    }

    override fun getFluidState(state: BlockState) =
        if (state[WATERLOGGED]) Fluids.WATER.getStill(false)
        else super.getFluidState(state)

    override fun getStateForNeighborUpdate(
        state: BlockState,
        direction: Direction?,
        neighborState: BlockState?,
        world: WorldAccess,
        pos: BlockPos,
        neighborPos: BlockPos?
    ): BlockState {
        return updateConnections(state, world, pos)
    }

    private fun updateConnections(state: BlockState, world: WorldAccess, pos: BlockPos): BlockState {
        val direction = state.get(FACING)
        val leftState = world.getBlockState(pos.offset(direction.rotateYClockwise()))
        val rightState = world.getBlockState(pos.offset(direction.rotateYCounterclockwise()))
        val frontState = world.getBlockState(pos.offset(direction))

        val connectedLeft = leftState.block is SofaBlock && (leftState[FACING] == direction || (leftState[FACING] == direction.rotateYCounterclockwise() && leftState[FRONT_CONNECTION] != FrontConnection.NONE))
        val connectedRight = rightState.block is SofaBlock && (rightState[FACING] == direction || (rightState[FACING] == direction.rotateYClockwise() && rightState[FRONT_CONNECTION] != FrontConnection.NONE))
        val connectedFront = frontState.block is SofaBlock
        val connectedFrontLeft = connectedFront && !connectedLeft && frontState[FACING] == direction.rotateYCounterclockwise()
        val connectedFrontRight = connectedFront && !connectedRight && frontState[FACING] == direction.rotateYClockwise()
        val frontConnection = when {
            connectedFrontLeft -> FrontConnection.LEFT
            connectedFrontRight -> FrontConnection.RIGHT
            else -> FrontConnection.NONE
        }

        return state
            .with(CONNECTED_LEFT, connectedLeft)
            .with(CONNECTED_RIGHT, connectedRight)
            .with(FRONT_CONNECTION, frontConnection)
    }

    override fun getOutlineShape(state: BlockState, view: BlockView?, pos: BlockPos?, context: ShapeContext?) =
        OUTLINE_SHAPE_MAP[
            Bits.buildSofaState(
                state[FACING],
                state[CONNECTED_LEFT],
                state[CONNECTED_RIGHT],
                state[FRONT_CONNECTION]
            )
        ]

    override fun getCollisionShape(state: BlockState, view: BlockView?, pos: BlockPos?, context: ShapeContext?) =
        COLLISION_SHAPE_MAP[
            Bits.buildSofaState(
                state[FACING],
                state[CONNECTED_LEFT],
                state[CONNECTED_RIGHT],
                state[FRONT_CONNECTION]
            )
        ]

    override fun mirror(state: BlockState, mirror: BlockMirror) =
        state.rotate(mirror.getRotation(state[FACING]))

    override fun rotate(state: BlockState, rotation: BlockRotation) =
        state.with(FACING, rotation.rotate(state[FACING]))

    override fun canPathfindThrough(state: BlockState, world: BlockView, pos: BlockPos, type: NavigationType) = false

    companion object {
        val FACING = Properties.HORIZONTAL_FACING
        val CONNECTED_LEFT = BooleanProperty.of("connected_left")
        val CONNECTED_RIGHT = BooleanProperty.of("connected_right")
        val FRONT_CONNECTION = EnumProperty.of("front", FrontConnection::class.java)
        val WATERLOGGED = Properties.WATERLOGGED

        private val OUTLINE_SHAPE_MAP: Byte2ObjectMap<VoxelShape>
        private val COLLISION_SHAPE_MAP: Byte2ObjectMap<VoxelShape>

        init {
            val bottom = createCuboidShape(0.0, 2.0, 0.0, 16.0, 7.0, 16.0)
            val leftArms = buildShapeRotations(5, 7, 13, 16, 13, 16)
            val rightArms = buildShapeRotations(5, 7, 0, 16, 13, 3)
            val thinLeftArms = buildShapeRotations(5, 7, 14, 16, 13, 16)
            val thinRightArms = buildShapeRotations(5, 7, 0, 16, 13, 2)
            val backs = buildShapeRotations(0, 7, 0, 5, 16, 16)
            val leftCorners = buildShapeRotations(5, 7, 11, 16, 16, 16)
            val rightCorners = buildShapeRotations(5, 7, 0, 16, 16, 5)
            val booleans = setOf(true, false)

            fun buildShapeMap(thin: Boolean): Byte2ObjectMap<VoxelShape> = Byte2ObjectOpenHashMap(
                Sets.cartesianProduct<Any>(FACING.values.toSet(), booleans, booleans, FRONT_CONNECTION.values.toSet()).map {
                    val facing = it[0] as Direction
                    val left = it[1] as Boolean
                    val right = it[2] as Boolean
                    val front = it[3] as FrontConnection

                    val parts = ArrayList<VoxelShape?>()
                    parts += backs[facing]

                    if (!left && front == FrontConnection.NONE) {
                        parts += if (thin) thinLeftArms[facing] else leftArms[facing]
                    }
                    if (!right && front == FrontConnection.NONE) {
                        parts += if (thin) thinRightArms[facing] else rightArms[facing]
                    }

                    when (front) {
                        FrontConnection.LEFT -> parts += leftCorners[facing]
                        FrontConnection.RIGHT -> parts += rightCorners[facing]
                        else -> {}
                    }

                    Bits.buildSofaState(facing, left, right, front) to VoxelShapes.union(
                        bottom,
                        *parts.filterNotNull().toTypedArray()
                    )
                }.toMap()
            )

            OUTLINE_SHAPE_MAP = buildShapeMap(thin = false)
            COLLISION_SHAPE_MAP = buildShapeMap(thin = true)
        }

        @JvmOverloads
        @JvmStatic
        fun getSleepingDirection(world: World, pos: BlockPos, ignoreNeighbors: Boolean = false): Direction? {
            val state = world.getBlockState(pos)
            if (state.block !is SofaBlock) return null

            val connectedLeft = state[CONNECTED_LEFT]
            val connectedRight = state[CONNECTED_RIGHT]
            val frontConnection = state[FRONT_CONNECTION]
            val facing = state[FACING]

            if ((!connectedLeft && !connectedRight && frontConnection == FrontConnection.NONE) || (!ignoreNeighbors && state[OCCUPIED]))
                return null

            val result = when {
                frontConnection != FrontConnection.NONE -> facing
                connectedLeft -> facing.rotateYClockwise()
                connectedRight -> facing.rotateYCounterclockwise()
                else -> null
            }

            if (result != null) {
                if (ignoreNeighbors) {
                    return result
                }
                val neighborState = world.getBlockState(pos.offset(result))
                if (neighborState.block is SofaBlock && !neighborState[OCCUPIED]) {
                    return result
                }
            }

            return null
        }
    }
}
